### PR TITLE
Render hooks

### PR DESCRIPTION
**Pull request by Cadiboo**
Added 3 hooks to allow Modders to control world rendering.

1) RenderBlockRenderLayerEvent - called from `RenderGlobal#renderBlockLayer(BlockRenderLayer, double, int, Entity)` & allows Modders to cancel the rendering of the BlockRenderLayer and/or inject their own rendering logic. Modders can inject their own rendering logic without canceling the event & let vanilla do it's normal rendering along with theirs.

2) RebuildChunkEvent - called from `RenderChunk#rebuildChunk(float, float, float, ChunkCompileTaskGenerator)` & allows Modders to cancel the rebuilding of the chunk and/or inject their own rebuilding logic. Modders can inject their own rebuilding logic without canceling the event & let vanilla do it's normal rebuilding along with theirs.

3) RebuildChunkBlocksEvent - called from `RenderChunk#rebuildChunk(float, float, float, ChunkCompileTaskGenerator)` (slightly after RebuildChunkEvent) & allows Modders to cancel the rebuilding of the chunk's blocks and/or inject their own block rebuilding logic. The difference between this and the RebuildChunkEvent is that if this event is cancelled TileEntities will still be rendered. Modders can inject their own block rebuilding logic without canceling the event & let vanilla do it's normal block rebuilding along with theirs.

Test mods for the 3 new events are included, however much, much more time & effort was spent on the RebuildChunkBlocksEvent.
Take a look at [this thread](http://www.minecraftforge.net/forum/topic/66516-1122-replace-world-renderer/) on the Forge Forums for more info (and possibilities for utilisations of these hooks)
Screenshots of what is possible with these events: (It took me a while to learn how to use the bufferbuilder in BLOCK mode)

### CODE REVIEW

**Comment by Cadiboo** [comment]
The patch CompiledChunk.java.patch can potentially be removed - all it does is change a protected method to public. This method could be used through reflection, however a patch seemed to me to be the most elegant solution.

File: patches/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
+--- ../src-base/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java
++++ ../src-work/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java
+@@ -14,7 +14,7 @@
+ {
+     public static final CompiledChunk field_178502_a = new CompiledChunk()
+     {
+-        protected void func_178486_a(BlockRenderLayer p_178486_1_)
++        public void func_178486_a(BlockRenderLayer p_178486_1_)
```
Comment (Gegy) [inline patches/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java.patch:8]: Should this be done with an AT?

Comment (Cadiboo) [reply patches/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java.patch:8]: Yes

**Comment by tterrag1098** [comment]
So with these events you are able to...screw up the render pipeline beyond all recognition? Anything else? Please provide a compelling reason for these rather invasive hooks.

**Comment by Cadiboo** [comment]
> So with these events you are able to...screw up the render pipeline beyond all recognition? Anything else? Please provide a compelling reason for these rather invasive hooks.

If you can recommend a way to add these hooks that doesn't mess with the render pipeline too much I'll rewrite the code to do it your way. I personally find the render pipeline code unreadable, an explanation of it (or a link to some documentation about it) would be greatly appreciated

With these hooks mods can add functionality (without breaking any other mods who add similar functionality) to do stuff like this:

**Comment by Cadiboo** [comment]
> So with these events you are able to...screw up the render pipeline beyond all recognition? Anything else? Please provide a compelling reason for these rather invasive hooks.

I don't see how any use of these events necessarily leads to anything different happening with the pipeline, 2 of the 3 examples I provided in the example mod still use BlockRenderDispatcher to render stuff, not touching the pipeline. The 3rd example could probably also have been done without bypassing the pipeline, It was just easier & more straightforward to write it using the BufferBuilder directly

**Comment by tterrag1098** [comment]
I was referring to all your "example" screenshots which just show extremely bugged out rendering and not anything useful. And screenshots of a 1.7 mod don't really show off what these hooks can do, since it's all theoretical.

**Comment by Cadiboo** [comment]
Uhhh... those “screenshots of a 1.7.10 mod” are screenshots I took playing my port of the mod to 1.12.2 using these new events. The only reason I didn’t include it in my PR is because it uses code I took from the old
mod and I’m not sure about the license on it. Here’s my repository with the code working 
[https://github.com/Cadiboo/MinecraftForge/commit/22f6ac8a25258e92d903d08bd06e05ba894176d7?files=1](https://github.com/Cadiboo/MinecraftForge/commit/22f6ac8a25258e92d903d08bd06e05ba894176d7?files=1)
Also in the screenshots I’m holding Diorite which was added in 1.8...

I believe that Better Foliage could also use these events rather than doing their core mod stuff.

**Comment by Cadiboo** [comment]
The original mod has 7+ files, with these events my port only needs about 50 lines of code in 1 file (and no ASM or reflection aside from an access transformer). However I haven’t made a GUI for it or allowed customisation of the smoothable blockstates, so that’s probably another 100 lines of code - but that’s not really relevant

**Comment by Cadiboo** [comment]
I haven’t checked Optifines code, but I think much of their rendering stuff (like joining glass panes) could be done with these events.

**Comment by Niko-sk2x** [comment]
Except optifine will never actively use any forge feature because it also has to work with vanilla.

**Comment by grondag** [comment]
These hooks don’t encourage compatibility or reuse. I think it is possible to provide rendering extension hooks that do  (it’s something I’m working on) but I dont think it belongs in Forge unless or until a de facto standard emerges.

Add: rendering hooks also need to be more granular. Most practical use cases won’t need to specially render everything in the event - just some parts or a certain aspect, like lighting. 

Trying to handle those distinctions in a cancellable event oversimplifies the problem.

**Comment by Cadiboo** [comment]
@grondag can I work with you (and hopefully the devs BetterFoliage) to make hooks that encourage comparability while still allowing mods like NoCubes to use them

**Comment by grondag** [comment]
@Cadiboo would be happy to work with you or anyone.  Will send you a DM. 

Note that NoCubes (smooth terrain) is already possible without any patches. Screen shots below show basalt and lava  from a soon-to-be-released volcano mod.  The block geometry is procedurally generated and baked and the cooled basalt can use the normal block rendering pipeline. The molten lava surface appearance is the only thing that requires a "render hook".  

TileEntitySpecialRenderer already allows this, but it doesn't scale well for something like terrain. Non-fast TESR is especially problematic because it requires authors to directly modify GL state and it's easy to mess up the whole frame that way. 

My approach is to create model extensions that allow each quad to specify a rendering pipeline using GLSL. The framework handles all the necessary plumbing and GL state wrangling. Authors get almost unlimited flexibility, performance remains solid,  and authors can only break their own stuff. 
   
I'd be happy to contribute any or all of this to Forge if there were an appetite but it needs more work and it's a _lot_ of code and very non-simple.  Could be a real maintenance headache.

File: patches/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java.patch (hunk 1 above)
Comment (Cadiboo) [reply patches/minecraft/net/minecraft/client/renderer/chunk/CompiledChunk.java.patch:8]: This actually can't be done with an AT
[http://www.minecraftforge.net/forum/topic/66980-1122-access-transformer-problem/](http://www.minecraftforge.net/forum/topic/66980-1122-access-transformer-problem/)

**Pull request closed without merging**